File: src/scripts/analyze_chimeric.py
#!/usr/bin/env python
from Bio import SeqIO
import pysam
import argparse
import numpy as np
import sys
import os

def calculate_identity(alignment):
    cigar = alignment.cigartuples
    if not cigar:
        return 0

    matches = sum(length for op, length in cigar if op == 7)  # '='
    mismatches = sum(length for op, length in cigar if op == 8)  # 'X'
    insertions = sum(length for op, length in cigar if op == 1)
    deletions = sum(length for op, length in cigar if op == 2)

    aligned_len = matches + mismatches # query_alignment_length is aligned_len + insertions
    total_bases = aligned_len + insertions + deletions
    pid = matches / total_bases if total_bases > 0 else 0.0
    return pid

# def read_bam(bam_file_path, reads_file_path):
def read_bam(bam_file_path, dot_file_path):
    # if (reads != ""):
    #     print(reads, flush=True)
    #     if reads == "-":
    #         handle = sys.stdin
    #     else:
    #         handle = open(reads, "r")

    #     read_lengths = {}
    #     for record in SeqIO.parse(handle, "fasta"):
    #         read_lengths[record.id] = len(record.seq)

    #     print("read", len(read_lengths), "read lengths", flush=True)

    node2len = {}
    with open(dot_file_path) as dot_file:
        line = dot_file.readline()
        while(line):
            if "label" not in line:
                line = dot_file.readline()
                continue
            if "->" in line:
                line = dot_file.readline()
                continue
            node = line.split()[0]
            if "+" not in node:
                node_len = line[line.find('L')+1: line.find('"', line.find('L')+1)]
                node2len[node] = int(node_len)
                # print("Length of " + node, node_len, flush=True)
            line = dot_file.readline()

    # Open the BAM file
    alignments = {}
    chimeric_edges = []
    with pysam.AlignmentFile(bam_file_path, "rb") as bam_file:
        # store alignments for each edge in the dict alignments
        for read in bam_file:
            if read.is_unmapped:
                continue
            # if read.reference_name != "84021.66_-85034.68":
            #     continue
            if calculate_identity(read) > 0.99:
                if read.reference_name not in alignments:
                    alignments[read.reference_name] = []
                alignments[read.reference_name].append({
                    "name": read.query_name,
                    "start": read.reference_start,
                    "end": read.reference_end,
                    "ref_length": bam_file.get_reference_length(read.reference_name),
                    "aligned": read.query_alignment_length,
                    "idt": calculate_identity(read),
                    # "aligned_fraction": read.query_alignment_length/read_lengths[read.query_name]
                })

        # detect chimeric at nodes
        tolerant_size = 1000
        # print("Contig", "Contig_len", "Node_start", "Node_end", "Read", "Aln_start", "Aln_end", "Aligned_len", "PI", "AF", sep="\t", flush=True)
        for r in alignments:
            sorted(alignments[r], key=lambda x: x["start"])

            node1, node2 = r.split('_')
            node1 = node1[:node1.find('.')]
            node2 = node2[:node2.find('.')]
            contig_len = bam_file.get_reference_length(r)

            # if r == "84021.66_-85034.68":
            #     for aln in alignments[r]:
            #         print(r, contig_len, node2len[node1], node2len[node2], aln["name"], aln["start"], aln["end"], aln["aligned"], aln["idt"], sep="\t", flush=True)

            split_coordinate1 = min(node2len[node1], contig_len - node2len[node2])
            split_coordinate2 = max(node2len[node1], contig_len - node2len[node2])
            if split_coordinate1 <= 2*tolerant_size or split_coordinate2 >= contig_len-2*tolerant_size:
                continue
            if node2len[node1] + node2len[node2] - contig_len >= 10000 - tolerant_size:
                continue

            cnt_f = 0
            cnt_r = 0
            cnt_all = 0
            for aln in alignments[r]:
                if aln["start"] <= max(split_coordinate1 - tolerant_size, 0) and aln["end"] > node2len[node1]:
                    cnt_f += 1
                if aln["start"] < contig_len - node2len[node2] and aln["end"] >= min(split_coordinate2 + tolerant_size, contig_len):
                    cnt_r += 1
                cnt_all += 1
            # print("contig name", r, "contig len", contig_len, "len 1", node2len[node1], "len 2", node2len[node2], "supporting reads", cnt_f, cnt_r, flush=True)
            if cnt_f <= 0 or cnt_r <= 0:
                print("contig name", r, "contig len", contig_len, "len 1", node2len[node1], "len 2", node2len[node2], "supporting reads", cnt_f, cnt_r, cnt_all, flush=True)
                print(r, "is chimeric", flush=True)
                chimeric_edges.extend(r.split('_'))

        # detect internal chimeric
        tolerant_size = 100
        for r in alignments:
            coverages = np.zeros(bam_file.get_reference_length(r), dtype=int)
            for i in alignments[r]:
                if i["end"]-tolerant_size > i["start"]:
                    coverages[i["start"]:i["end"]-tolerant_size] += 1
            
            potential_pos = []
            if np.average(coverages) >= 5:
                for i in range(bam_file.get_reference_length(r)):
                    if coverages[i] <= 1 and i >= 5001- tolerant_size and i <= bam_file.get_reference_length(r)-5001+tolerant_size:
                        potential_pos.append(i)
                        # print("low coverage:", r, i)
            
            reads = set()
            for i in alignments[r]:
                for j in potential_pos:
                    if i["start"] <= j-5001+tolerant_size and i["end"] >= j+5001-tolerant_size:
                        reads.add(i["name"])
            
            # print IDs of chimeric reads
            if len(reads) != 0:
                print(r, "is chimeric (internal)", flush=True)
                chimeric_edges.extend(r.split('_'))
    return chimeric_edges
            
if __name__ == "__main__":
    # Set up argument parsing
    parser = argparse.ArgumentParser(description="Read a BAM file and print alignment details.")
    parser.add_argument("bam_file", help="Path to the input BAM file")
    parser.add_argument("dot_file", help="Path to the input dot file (for node sizes)")
    parser.add_argument("output", help="Path to the output file")
    # parser.add_argument("reads", help="Path to reads", default="", required=False)
    
    # Parse the command-line arguments
    args = parser.parse_args()

    # Call the function with the BAM file path provided by the user
    # read_bam(args.bam_file, args.reads_file)
    # chimeric_edges = read_bam(args.bam_file, args.dot_file, args.reads)
    if not os.path.isfile(args.output):
        chimeric_edges = read_bam(args.bam_file, args.dot_file)
        with open(args.output, "w") as fout:
            for e in chimeric_edges:
                fout.write(e + '\n')

File: src/scripts/analyze_chimeric.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>
#include <vector>
#include <set>
#include <htslib/sam.h>
#include <cassert>
#include <algorithm>
#include <numeric>

// Calculate percent identity from a BAM alignment
double calculate_identity(const bam1_t* aln) {
    uint32_t* cigar = bam_get_cigar(aln);
    int n_cigar = aln->core.n_cigar;
    int matches = 0, mismatches = 0, insertions = 0, deletions = 0;
    for (int i = 0; i < n_cigar; ++i) {
        int op = bam_cigar_op(cigar[i]);
        int len = bam_cigar_oplen(cigar[i]);
        if (op == BAM_CMATCH) matches += len;
        else if (op == BAM_CEQUAL) matches += len;
        else if (op == BAM_CDIFF) mismatches += len;
        else if (op == BAM_CINS) insertions += len;
        else if (op == BAM_CDEL) deletions += len;
    }
    int aligned_len = matches + mismatches;
    int total_bases = aligned_len + insertions + deletions;
    return total_bases > 0 ? (double)matches / total_bases : 0.0;
}

// Parse node lengths from DOT file
std::unordered_map<std::string, int> parse_dot_file(const std::string& dot_file_path) {
    std::unordered_map<std::string, int> node2len;
    std::ifstream dot_file(dot_file_path);
    std::string line;
    while (std::getline(dot_file, line)) {
        if (line.find("label") == std::string::npos) continue;
        if (line.find("->") != std::string::npos) continue;
        std::string node = line.substr(0, line.find(' '));
        if (node.find('+') == std::string::npos) {
            size_t lpos = line.find('L');
            size_t qpos = line.find('"', lpos + 1);
            std::string node_len = line.substr(lpos + 1, qpos - lpos - 1);
            node2len[node] = std::stoi(node_len);
        }
    }
    return node2len;
}

int main(int argc, char* argv[]) {
    if (argc < 4) {
        std::cerr << "Usage: " << argv[0] << " <bam_file> <dot_file> <output_file>\n";
        return 1;
    }
    std::string bam_file_path = argv[1];
    std::string dot_file_path = argv[2];
    std::string output_file = argv[3];

    auto node2len = parse_dot_file(dot_file_path);

    samFile* fp = sam_open(bam_file_path.c_str(), "r");
    bam_hdr_t* hdr = sam_hdr_read(fp);
    bam1_t* aln = bam_init1();

    std::unordered_map<std::string, std::vector<std::tuple<std::string, int, int, int, int, double>>> alignments;
    std::set<std::string> chimeric_edges;

    while (sam_read1(fp, hdr, aln) >= 0) {
        if (aln->core.flag & BAM_FUNMAP) continue;
        std::string ref_name = hdr->target_name[aln->core.tid];
        double idt = calculate_identity(aln);
        if (idt > 0.99) {
            alignments[ref_name].emplace_back(
                bam_get_qname(aln),
                aln->core.pos,
                bam_endpos(aln),
                hdr->target_len[aln->core.tid],
                aln->core.l_qseq,
                idt
            );
        }
    }

    // Chimeric detection logic (simplified, adapt as needed)
    int tolerant_size = 1000;
    for (const auto& kv : alignments) {
        const std::string& r = kv.first;
        const auto& alns = kv.second;
        std::vector<std::string> nodes;
        size_t uscore = r.find('_');
        if (uscore == std::string::npos) continue;
        nodes.push_back(r.substr(0, uscore));
        nodes.push_back(r.substr(uscore + 1));
        std::string node1 = nodes[0].substr(0, nodes[0].find('.'));
        std::string node2 = nodes[1].substr(0, nodes[1].find('.'));
        int contig_len = hdr->target_len[sam_hdr_name2tid(hdr, r.c_str())];
        int split_coordinate1 = std::min(node2len[node1], contig_len - node2len[node2]);
        int split_coordinate2 = std::max(node2len[node1], contig_len - node2len[node2]);
        if (split_coordinate1 <= 2 * tolerant_size || split_coordinate2 >= contig_len - 2 * tolerant_size) continue;
        if (node2len[node1] + node2len[node2] - contig_len >= 10000 - tolerant_size) continue;

        int cnt_f = 0, cnt_r = 0, cnt_all = 0;
        for (const auto& aln_tuple : alns) {
            int start = std::get<1>(aln_tuple);
            int end = std::get<2>(aln_tuple);
            if (start <= std::max(split_coordinate1 - tolerant_size, 0) && end > node2len[node1]) cnt_f++;
            if (start < contig_len - node2len[node2] && end >= std::min(split_coordinate2 + tolerant_size, contig_len)) cnt_r++;
            cnt_all++;
        }
        if (cnt_f <= 0 || cnt_r <= 0) {
            std::cout << "contig name " << r << " contig len " << contig_len << " len 1 " << node2len[node1] << " len 2 " << node2len[node2]
                << " supporting reads " << cnt_f << " " << cnt_r << " " << cnt_all << std::endl;
            std::cout << r << " is chimeric" << std::endl;
            chimeric_edges.insert(node1);
            chimeric_edges.insert(node2);
        }
    }

    // Internal chimeric detection
    tolerant_size = 100;
    for (const auto& kv : alignments) {
        const std::string& r = kv.first;
        int ref_len = hdr->target_len[sam_hdr_name2tid(hdr, r.c_str())];
        std::vector<int> coverages(ref_len, 0);

        // Build coverage array
        for (const auto& aln_tuple : kv.second) {
            int start = std::get<1>(aln_tuple);
            int end = std::get<2>(aln_tuple);
            if (end - tolerant_size > start) {
                int cov_start = std::max(start, 0);
                int cov_end = std::max(end - tolerant_size, cov_start);
                for (int i = cov_start; i < cov_end && i < ref_len; ++i) {
                    coverages[i]++;
                }
            }
        }

        // Find potential positions with low coverage
        std::vector<int> potential_pos;
        double avg_cov = std::accumulate(coverages.begin(), coverages.end(), 0.0) / ref_len;
        if (avg_cov >= 5) {
            for (int i = 0; i < ref_len; ++i) {
                if (coverages[i] <= 1 &&
                    i >= 5001 - tolerant_size &&
                    i <= ref_len - 5001 + tolerant_size) {
                    potential_pos.push_back(i);
                }
            }
        }

        // Find reads supporting chimeric positions
        std::set<std::string> reads;
        for (const auto& aln_tuple : kv.second) {
            int aln_start = std::get<1>(aln_tuple);
            int aln_end = std::get<2>(aln_tuple);
            std::string aln_name = std::get<0>(aln_tuple);
            for (int j : potential_pos) {
                if (aln_start <= j - 5001 + tolerant_size &&
                    aln_end >= j + 5001 - tolerant_size) {
                    reads.insert(aln_name);
                }
            }
        }

        // Print IDs of chimeric reads
        if (!reads.empty()) {
            std::cout << r << " is chimeric (internal)" << std::endl;
            size_t uscore = r.find('_');
            if (uscore != std::string::npos) {
                chimeric_edges.insert(r.substr(0, uscore));
                chimeric_edges.insert(r.substr(uscore + 1));
            }
        }
    }

    // Output chimeric edges
    std::ofstream fout(output_file);
    for (const auto& e : chimeric_edges) {
        fout << e << std::endl;
    }

    bam_destroy1(aln);
    bam_hdr_destroy(hdr);
    sam_close(fp);
    return 0;
}